### PR TITLE
Segmented signing checks on extensions

### PR DESCRIPTION
**Pull request by carlopi**
Bottleneck of sha256 computation is alleviated by computing sha256 on fixed-size segments, and then combining the hashes. This is basically a 2-level Merkle tree, with root, no internal nodes, and file_size / 1MB leafs.

Note that 2-level sha256 computation is computed both on threaded and non-threaded versions, since results will have to agree (and in any case the second step has constant negligible cost).

### CODE REVIEW

File: src/main/extension/extension_load.cpp [hunk 1]
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
...
+		std::vector<idx_t> splits(numChunks + 1);
+
+		splits.back() = signature_offset;
+		for (idx_t i = 0; i < chunks.size(); i++)
+			splits[i] = maxLenChunks * i;
+
+#ifndef DUCKDB_NO_THREADS
+		std::deque<std::thread> threads;
+		for (idx_t i = 0; i < numChunks; i++)
+			threads.push_back(std::thread(computeSHA256, handle.get(), splits[i], splits[i+1], &chunks[i]));
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:99]: nit: can we use braces on all of these loops as per the coding guidelines?

File: src/main/extension/extension_load.cpp [hunk 2]
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
 
 		auto signature_offset = handle->GetFileSize() - signature.size();
 
+		const idx_t maxLenChunks = 1024 * 1024;
+		const idx_t numChunks = (signature_offset - 1) / maxLenChunks;
+		std::vector<std::string> chunks(numChunks);
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:89]: What happens if the extension is smaller than `maxLenChunks`?

File: src/main/extension/extension_load.cpp [hunk 3]
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
...
+		for (idx_t i = 0; i < numChunks; i++)
+			computeSHA256(handle.get(), splits[i], splits[i + 1], &chunks[i]);
+#endif // DUCKDB_NO_THREADS
+
 		string file_content;
-		file_content.resize(signature_offset);
-		handle->Read((void *)file_content.data(), signature_offset, 0);
+		file_content.resize(256 * numChunks);
+		for (idx_t i = 0; i < numChunks; i++) {
+			for (idx_t j = 0; j < 256; j++)
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:111]: Can we use a `memcpy` here instead, and move `256` to a named constexpr?

File: src/main/extension/extension_load.cpp [hunk 4]
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
...
+		const idx_t numChunks = (signature_offset - 1) / maxLenChunks;
+		std::vector<std::string> chunks(numChunks);
+		std::vector<idx_t> splits(numChunks + 1);
+
+		splits.back() = signature_offset;
+		for (idx_t i = 0; i < chunks.size(); i++)
+			splits[i] = maxLenChunks * i;
+
+#ifndef DUCKDB_NO_THREADS
+		std::deque<std::thread> threads;
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:97]: Can this not be `vector`?

File: src/main/extension/extension_load.cpp [hunk 5]
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
...
+		for (idx_t i = 0; i < numChunks; i++)
+			threads.push_back(std::thread(computeSHA256, handle.get(), splits[i], splits[i+1], &chunks[i]));
+
+		for (idx_t i = 0; i < numChunks; i++)
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:101]: `for(auto &thread : threads)`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. I'm not a big fan of launching threads directly in the `TryInitialLoad` - but as extensions can be loaded before the task scheduler exists (when loaded during start-up) perhaps it is the best/only way.

Some other comments:

**Comment by carlopi** [comment]
Thanks for the various feedback!

> I'm not a big fan of launching threads directly in the TryInitialLoad - but as extensions can be loaded before the task scheduler exists (when loaded during start-up) perhaps it is the best/only way.

On threads, was very much unsure of where to go, and left it as optional commit on purpose (so that a different solution could also be rolled in).
Manual handling looked [to me] simpler here, at the very least an explanation why it went against codebase standard and your comment that task scheduler might not be there should be added.

I will fix the comments (thanks!!), check signign actually works on my fork, and then I will ping back.

File: src/main/extension/extension_load.cpp (hunk 3 above)
Comment (carlopi) [reply src/main/extension/extension_load.cpp]: Moved to reserve length + string concatenation, that should be the same but seems clearer.

**Comment by samansmink** [comment]
Good idea! LGTM!

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**